### DISCUSSION TITLE
Please help with Native Library

### DISCUSSION THREAD

**Discussion by AlexeyBKuznetsov** [dotnet/maui#6872] [Q&A]
Hello,

I am using MAUI to build the cross-platform application and want to use existing c++ native library already built per platform

So, here is the setup I have (attaching the Visual Studio test solution)
1. Pre-built NativeLib 
  i. NativeLib.Windows.x64.dll
  ii. NativeLib.Android.arm64.so
  iii. NativeLib.iOS.arm64.dylib

2. NativeLibTest.Import - .NET library to
  i. Reference NativeLib(s) above
  ii. Define the DllImport(s)
  iii. Wrap DllImport(s) into .NET style API

3. NativeLibTest.MAUI - sample MAUI application which uses NativeLibTest.Import

Now, here are a few questions I need help please :
1. I am adding NativeLib.Windows.x64.dll to the NativeLibTest.Import.csproj and it is later properly copied to NativeLibTest.MAUI Windows App during the deployment and everything works fine

2. However, when I am adding NativeLib.Android.arm64.so in a similar way, it is not copied to NativeLibTest.MAUI Android App during the deployment. Another option is to add by using <AndroidNativeLibrary>, but this doesn't seem to work in .NET library project

3. So, for Android I have to add NativeLib.Android.arm64.so to the NativeLibTest.MAUI.csproj and in such case NativeLib.Android.arm64.so properly copied during the deployment and works. But, this is very inconvenient as I want to have NativeLib.Android.arm64.so on the NativeLibTest.Import library level and not on the application level

4. For iOS, I am trying approach similar to Windows, but it doesn’t work. I am also trying to add NativeLib.iOS.arm64.dylib by <NativeReference> in the NativeLibTest.Import.csproj and NativeLibTest.MAUI.csproj, but doesn’t work

5. The suggested topic https://docs.microsoft.com/en-us/xamarin/ios/platform/native-interop says to use MonoTouch.ObjCRuntime.Dlfcn.dlopen ("/full/path/to/Animal.dylib", 0); but I am not exactly sure how to use this in the NativeLibTest.Import library project

Adding Visual Studio test solution :
[NativeLibTest.zip](https://github.com/dotnet/maui/files/8635531/NativeLibTest.zip)

Please help
Thanks

**Comment by xtuzy** [comment]
I separate .dll and .so to different library, see https://github.com/xtuzy/ApplePDF/tree/ApplePDF, if use single project manage they, maybe need use `<PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|AnyCPU' and '$(TargetFramework)' == 'net6.0-android'">` to combine two .csproj to one .csproj

**Comment by AlexeyBKuznetsov** [comment]
Thanks xtuzy,

I tried couple approaches based on your example, but with no success
1. I tried loading libtest_lib.dylib via dlopen() - I checked file is available, however system is not able to load it and returns zero handle. I was doing on new iPad Pro and libtest_lib.dylib is compiled for arm64. Any ideas please ?

2. I create iOS Binding project and conditionally added as a project reference to the MAUI app.
```
<ItemGroup>
<ProjectReference Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'ios'" Include="..\iOSBindings\iOSBindings.csproj" />
</ItemGroup>
```
I received next build errors
```
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0). Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-arm. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-arm64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-x64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-x86. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / ios-arm64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / maccatalyst-x64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / win10-x64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
```
Could you please suggest what am I doing wrong ?

	
Also, attaching libtest_lib.dylib in case you would have time to please take a look and suggest if anything is wrong with the library
[libtest_lib.zip](https://github.com/dotnet/maui/files/8710516/libtest_lib.zip)

Thanks

**Comment by xtuzy** [comment]
you are using a dynamic library , apple store not accept it , if you have a static library,  you can see a best example at https://docs.microsoft.com/en-us/xamarin/cross-platform/cpp/
souce code at https://github.com/xamcat/mobcat-samples/tree/master/cpp_with_xamarin .

**Comment by AlexeyBKuznetsov** [comment]
Thanks xtuzy for the suggestion, I saw this great tutorial. The only difference : he creates iOS binding project and packs to Nuget package and later reference the Nuget package. What I am trying to do is to create iOS binding project and conditionally reference to the MAUI application and receive the errors I indicated above
```
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0). Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-arm. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-arm64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-x64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / android-x86. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / ios-arm64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / maccatalyst-x64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
MauiApp1.csproj : error NU1201: Project NativeLibrary is not compatible with net6.0-ios15.4 (.NETCoreApp,Version=v6.0) / win10-x64. Project NativeLibrary supports: xamarinios10 (Xamarin.iOS,Version=v1.0)
```

This is where I need help - the issue is very simple to reproduce
1. create iOS binding project
2. create MAUI project
3. conditionally reference iOS binding project to MAUI project

Any suggestions please ?
Thanks

**Reply by xtuzy** [reply]
If you have .a, you see my project structure .
binding library: https://github.com/xtuzy/ApplePDF/blob/ApplePDF/ApplePDF.NativeAssets.iOS/ApplePDF.NativeAssets.net6iOS.csproj. 
maui app: https://github.com/xtuzy/ApplePDF/blob/ApplePDF/ApplePDF.Demo.Maui/ApplePDF.Demo.Maui.csproj
i try to use you .dylib, but fail, you need ask maui team

**Comment by AlexeyBKuznetsov** [comment]
Thanks xtuzy, I am following the steps from your project, however still not able to make it working on the iOS. Attaching simple solution which has Shared project (NativeLib.Shared), projects per Platform, combined (NativeLib) and consumer (NativeLib.Maui) where I call NativeLib inside MainPage.xaml.cs. Could you please take a look. Your help is much appreciated, thanks !
[NativeLibTest.zip](https://github.com/dotnet/maui/files/8774512/NativeLibTest.zip)

**Comment by AlexeyBKuznetsov** [comment]
As always, thanks a lot xtuzy! Unfortunately, it still not working for my iPad. I keep looking for the solution, thanks anyway

**Reply by Berginator55** [reply]
I’ve been looking for several months for an example where someone is actually interfacing with a native c/c++ library from Maui (for an android platform for now).  There is a simplistic Xamarin example that passes an argument by value and returns via the native function’s return argument. I was not having any luck getting this rather useless test case to actually to work in Maui regardless, so what you've done with your test example here helped me a little at least with the semantics. Have you had any further success with your efforts? 

I need to pass a byte array to a native c/c++ library and have it return a byte array. This has been frustrating navigating the Maui semantics with little to no info or help coming from MS. It is as if they want us to abandon established validated codes and re-write everything in MS's latest next best thing since their last best thing every few years. :(

**Comment by stevexuzeta** [comment]
Is there any solution? I hope to find a example and a guide.

**Reply by xtuzy** [reply]
android, windows, ios : https://github.com/xtuzy/Veldrid.SPIRV.NativeAsset
android, windows: https://github.com/xtuzy/ApplePDF
ios, mac: https://github.com/xtuzy/Veldrid.VK.NativeAsset

i feel binding c/c++ library is easy, binding library have the same format
